File: src/main/java/com/phasmidsoftware/dsaipg/sort/linearithmic/QuickSort_DualPivot.java
/*
 * Copyright (c) 2024. Robin Hillyard
 */

 package com.phasmidsoftware.dsaipg.sort.linearithmic;

 import com.phasmidsoftware.dsaipg.sort.Helper;
 import com.phasmidsoftware.dsaipg.sort.SortException;
 import com.phasmidsoftware.dsaipg.util.Config;
 
 import java.util.ArrayList;
 import java.util.List;
 
 import static com.phasmidsoftware.dsaipg.sort.InstrumentedComparatorHelper.getRunsConfig;
 
 /**
  * Class QuickSort_DualPivot which extends QuickSort.
  *
  * @param <X> the underlying comparable type.
  */
 public class QuickSort_DualPivot<X extends Comparable<X>> extends QuickSort<X> {
 
     public static final String DESCRIPTION = "QuickSort dual pivot";
 
     public QuickSort_DualPivot(String description, int N, int nRuns, Config config) {
         super(description, N, nRuns, config);
         setPartitioner(createPartitioner());
     }
 
     /**
      * Constructor for QuickSort_DualPivot
      *
      * @param helper an explicit instance of Helper to be used.
      */
     public QuickSort_DualPivot(Helper<X> helper) {
         super(helper);
         setPartitioner(createPartitioner());
     }
 
     /**
      * Constructor for QuickSort_DualPivot
      *
      * @param N      the number elements we expect to sort.
      * @param nRuns  the number of runs.
      * @param config the configuration.
      */
     public QuickSort_DualPivot(int N, int nRuns, Config config) {
         this(DESCRIPTION, N, nRuns, config);
     }
 
     /**
      * Constructor for QuickSort_DualPivot
      *
      * @param N      the number elements we expect to sort.
      * @param config the configuration.
      */
     public QuickSort_DualPivot(int N, Config config) {
         this(DESCRIPTION, N, getRunsConfig(config), config);
     }
 
     public Partitioner<X> createPartitioner() {
         return new Partitioner_DualPivot(getHelper());
     }
 
     public class Partitioner_DualPivot implements Partitioner<X> {
 
         public Partitioner_DualPivot(Helper<X> helper) {
             this.helper = helper;
         }
 
         /**
          * Method to partition the given partition into smaller partitions.
          *
          * @param partition the partition to divide up.
          * @return a list of partitions, whose length depends on the sorting method being used.
          */
         public List<Partition<X>> partition(Partition<X> partition) {
             int n = partition.to - partition.from;
             if (n < 3) throw new SortException("cannot use DualPivot partitioning when size is less than 3");
             final X[] xs = partition.xs;
             final int p1 = partition.from;
             final int p2 = partition.to - 1;
             helper.swapConditional(xs, p1, p2);  // Ensure p1 and p2 are in correct order
             int lt = p1 + 1;
             int gt = p2 - 1;
             int i = lt;
             X v1 = xs[p1];
             X v2 = xs[p2];
 
             while (i <= gt) {
                 X x = xs[i];
                 if (x.compareTo(v1) < 0) {
                     swap(xs, lt++, i++);
                 } else if (x.compareTo(v2) > 0) {
                     swap(xs, i, gt--);
                 } else {
                     i++;
                 }
             }
 
             swap(xs, p1, --lt);
             swap(xs, p2, ++gt);
 
             List<Partition<X>> partitions = new ArrayList<>();
             partitions.add(new Partition<>(xs, p1, lt));
             partitions.add(new Partition<>(xs, lt + 1, gt));
             partitions.add(new Partition<>(xs, gt + 1, p2 + 1));
             return partitions;
         }
 
         // Swap utility method
         private void swap(X[] ys, int i, int j) {
             X temp = ys[i];
             ys[i] = ys[j];
             ys[j] = temp;
         }
 
         private final Helper<X> helper;
     }
 }
 

File: Python/src/sort/linearithmic/quick_sort_dual_pivot.py
"""
Dual-pivot quicksort, ported from
sort/linearithmic/QuickSort_DualPivot.java.

Two pivots split the range into three parts rather than two, so the recursion is
shallower and each pass does more. This is the algorithm Java's own
Arrays.sort uses for primitives.
"""

from __future__ import annotations

from typing import TypeVar

from src.sort.generic.sort_exception import SortException
from src.sort.helper.helper import Helper
from src.sort.linearithmic.partition import Partition, Partitioner
from src.sort.linearithmic.quick_sort import QuickSort

X = TypeVar("X")

DESCRIPTION = "QuickSort dual pivot"


class PartitionerDualPivot(Partitioner[X]):
    """
    Partition about two pivots, into the elements below the first, those between
    the two, and those above the second.
    """

    def __init__(self, helper: Helper[X]) -> None:
        """
        :param helper: the Helper to compare and exchange through.
        """
        self.helper = helper

    def partition(self, partition: Partition[X]) -> list[Partition[X]]:
        """
        Split the range about two pivots.

        :param partition: the range to split.
        :return: the three ranges between and either side of the pivots.
        :raises SortException: if the range holds fewer than three elements.
        """
        helper = self.helper
        xs = partition.xs
        n = partition.to - partition.from_
        if n < 3:
            raise SortException("cannot use DualPivot partitioning when size is less than 3")
        p1 = partition.from_
        p2 = partition.to - 1
        # ensure the smaller pivot is at p1
        helper.swap_conditional(xs, p1, p2)
        lt = p1 + 1
        gt = p2 - 1
        i = lt
        v1 = xs[p1]
        v2 = xs[p2]
        x_lt = helper.get(xs, lt)
        x_gt = helper.get(xs, gt)
        xi = x_lt
        while i <= gt:
            if helper.compare(helper.lookup(xi), v1) < 0:
                helper.swap_vw(x_lt, xi, xs, lt, i)
                lt += 1
                i += 1
                xi = helper.get(xs, i)
                x_lt = helper.get(xs, lt) if lt != i else xi
            elif helper.compare(xi, v2) > 0:
                helper.swap_vw(xi, x_gt, xs, i, gt)
                gt -= 1
                if i == lt:
                    x_lt = x_gt
                xi = x_gt
                x_gt = helper.get(xs, gt)
            else:
                i += 1
                xi = helper.get(xs, i)
        # NOTE these restore the two pivots to their final places, and either can
        # be a no-op: lt is still p1 + 1 if nothing was less than v1, and gt is
        # still p2 - 1 if nothing was greater than v2, which happens readily when
        # the input has many duplicates. helper.swap asserts i != j, because
        # counting a self-swap would corrupt the instrumentation, so skip the
        # swap rather than ask for one.
        lt -= 1
        if lt != p1:
            helper.swap(xs, p1, lt)
        gt += 1
        if gt != p2:
            helper.swap(xs, p2, gt)
        return [
            Partition(xs, partition.from_, lt),
            Partition(xs, lt + 1, gt),
            Partition(xs, gt + 1, partition.to),
        ]


class QuickSortDualPivot(QuickSort[X]):
    """
    Quicksort with two pivots.
    """

    def create_partitioner(self) -> Partitioner[X]:
        return PartitionerDualPivot(self.get_helper())
